mautrix_telegram/util/time_func.py

```diff
@@ -0,0 +1,33 @@
+from asyncio import Future, isfuture
+from functools import wraps
+from time import time
+from types import FunctionType
+from inspect import isawaitable
+import logging
+from typing import Optional
+
+log = logging.getLogger("mau.time_func")
+
+def time_func(log_level: str, name: Optional[str] = None):
+    def decorator(func: FunctionType):
+        @wraps(func)
+        def wrapper_function(*args, **kwargs):
+            int_level = logging.getLevelName(log_level)
+            func_name = name or func.__name__
+            t1 = time()
+            res = func(*args,  **kwargs)
+            if isawaitable(res):
+                # Ensure we only log once the function completes
+                async def await_result():
+                    t2 = time()
+                    awaited_res = await res
+                    invocation_time = time() - t1
+                    await_time = time() - t2
+                    log.log(int_level, f"async func {func_name} took {round(invocation_time+await_time, 3)}s (invoke {round(invocation_time, 3)}s, await {round(await_time, 3)})")
+                    return awaited_res
+                return await_result()
```

Comment:
This changes the return type of the function from a Future/Coroutine/Task to a Coroutine.

I am unsure if this matters in reality, but obviously if any function depended on this then it would break. We could instead make this only work for coroutines, and then update any functions that use this to be coroutines?